README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
ARHT
====

Perform the Adaptable Regularized Hotelling's *T*<sup>2</sup> test (ARHT) proposed by Li et al., (2016) arXiv:1609.08725. Both one- and two- sample mean test are available with various probabilistic alternative prior models. It contains a function to consistently estimate higher order moments of the population covariance spectral distribution using the spectral of the sample covariance matrix (Bai et al. (2010) <doi:10.1111/j.1467-842X.2010.00590.x>). In addition, it contains a function to sample from 3-variate chi-squared random vectors approximately with a given correlation matrix when the degrees of freedom are large.

Installation
------------

You can install ARHT from github with:

``` r
# install.packages("devtools")
devtools::install_github("HaoranLi/ARHT")
```

Example
-------

This is a basic example which shows you how to solve a common problem:

``` r
library(ARHT)
## basic example code
set.seed(10086)
# One-sample test
n1 = 300; p =500
dataX = matrix(rnorm(n1 * p), nrow = n1, ncol = p)
res1 = ARHT(dataX)

# Two-sample test
n2= 400
dataY = matrix(rnorm(n2 * p), nrow = n2, ncol = p )
res2 = ARHT(dataX, dataY, mu_0 = rep(0.01,p))

# Specify probabilistic alternative priors model
res3 = ARHT(dataX, dataY, mu_0 = rep(0.01,p), 
            prob_alt_prior = list(c(1/3, 1/3, 1/3), c(0,1,0)))
# Change Type 1 error calibration method
res4 = ARHT(dataX, dataY, mu_0 = rep(0.01,p),
            Type1error_calib = "sqrt")
RejectOrNot = res4$ARHT_pvalue < 0.05
```


NEWS.md
News
================
Haoran Li
2/27/2018

ARHT 1.0
========

This is the initial submission of ARHT.


R/r3chisq.R
#' 3-variate positively correlated chi-squared sample generation when degrees of freedom are large
#'@description Generate samples approximately from three positively correlated chi-squared random variables
#' \eqn{(\chi^2(d_1), \chi^2(d_2), \chi^2(d_3))}
#' when the degrees of freedom \eqn{(d_1, d_2, d_3)} are large.
#' @name r3chisq
#' @details It is generally hard to sample from \eqn{(\chi^2(d_1), \chi^2(d_2), \chi^2(d_3))} with a designed correlation matrix.
#'  In the algorithm, we approximate the random vector by \eqn{(z^T Q_1 z, z^T Q_2 z, z^T Q_3 z)}
#'  where \eqn{z} is a standard norm random vector and \eqn{Q_1,Q_2,Q_3} are diagonal matrices
#'  with diagonal elements 1's and 0's. The designed positive correlations is approximated by carefully
#'  selecting common locations of 1's on the diagonals. The generated sample may have slightly larger marginal degrees
#'  of freedom than the inputted \code{df}, also slightly different covariances.
#' @param size sample size.
#' @param df the degree of freedoms of the marginal distributions. Must be non-negative, but can be non-integer.
#'  The function uses \code{ceiling(df)} if non-integer.
#' @param corr_mat the target correlation matrix; negative elements will be set to 0.
#' @return
#' \itemize{
#' \item{\code{sample}}: a \code{size}-by-3 matrix contains the generated sample.
#' \item{\code{approx_cov}}: the true covariance matrix of \code{sample}.}
#' @references Li, H., Aue, A., Paul, D., Peng, J., & Wang, P. (2016). \emph{
#' An adaptable generalization of Hotelling's  \eqn{T^2}
#'  test in high dimension.} arXiv preprint <arXiv:1609.08725>.
#' @examples
#' set.seed(10086)
#' cor_examp = matrix(c(1,1/6,2/3,1/6,1,2/3,2/3,2/3,1),3,3)
#' a_sam = r3chisq(size = 10000,
#'                 df = c(80,90,100),
#'                 corr_mat = cor_examp)
#' cov(a_sam$sample) - a_sam$approx_cov
#' cov2cor(a_sam$approx_cov) - cor_examp
#' @export
r3chisq = function(size, df, corr_mat){

        if(!is.numeric(size)){
                stop("size must be numeric")
        }
        if(length(size) != 1L){
                stop("length(size) must be 1")
        }
        if(!is.vector(df, mode = "numeric")){
                stop("df must be a numeric vector")
        }
        if(length(df) != 3L){
                stop("length(df) must be 3")
        }
        if(any(df < 0)){
                stop("df must be positive")
        }
        df = ceiling(df)

        if( (length(dim(corr_mat)) > 2L) || !(is.numeric(corr_mat))){
                stop("corr_mat must be a numeric matrix")
        }
        if(!is.matrix(corr_mat)){
                corr_mat = as.matrix(corr_mat)
        }
        if(!isSymmetric(corr_mat)){
                stop("corr_mat must be symmetric")
        }
        corr_mat = corr_mat * (corr_mat >= 0)

        half_covariances = floor(c((df[1])^(1/2) * (df[2])^(1/2) * corr_mat[1,2],
                                   (df[1])^(1/2) * (df[3])^(1/2) * corr_mat[1,3],
                                   (df[2])^(1/2) * (df[3])^(1/2) * corr_mat[2,3]))

        # Build three diag matrices Q1, Q2, Q3.  z^TQ_1z, z^TQ_2z, z^TQ_3z would be the chi-square vector
        # See supplementary material for algorithm explanation
        mincov = min(half_covariances)
        part1 = c(rep(1, mincov), rep(1, half_covariances[1]-mincov), rep(1, half_covariances[2]-mincov), rep(0, half_covariances[3]-mincov))
        part2 = c(rep(1, mincov), rep(1, half_covariances[1]-mincov), rep(0, half_covariances[2]-mincov), rep(1, half_covariances[3]-mincov))
        part3 = c(rep(1, mincov), rep(0, half_covariances[1]-mincov), rep(1, half_covariances[2]-mincov), rep(1, half_covariances[3]-mincov))
        ramainder1 = max(df[1] - sum(part1), 0)
        ramainder2 = max(df[2] - sum(part2), 0)
        ramainder3 = max(df[3] - sum(part3), 0)
        Q1 = c(part1, rep(1, ramainder1), rep(0, ramainder2), rep(0, ramainder3))
        Q2 = c(part2, rep(0, ramainder1), rep(1, ramainder2), rep(0, ramainder3))
        Q3 = c(part3, rep(0, ramainder1), rep(0, ramainder2), rep(1, ramainder3))
        Qs = cbind(Q1, Q2, Q3)

        #chi-squared(1) sample
        bootstrap_sample = matrix((rnorm(length(Q1) * size))^2, nrow = size, ncol = length(Q1))
        chisq = bootstrap_sample %*% Qs
        approx_cov = 2* matrix(c(sum(Q1), half_covariances[1], half_covariances[2],
                                 half_covariances[1], sum(Q2), half_covariances[3],
                                 half_covariances[2], half_covariances[3], sum(Q3)),
                                nrow = 3, ncol = 3)
        return(list(sample = chisq, approx_cov = approx_cov))
}




R/moments_PSD.R
#' Consistent estimators of high-order moments of the population spectral distribution for high-dimensional data
#' @description  The function calculates consistent estimators of moments of the spectral distribution
#' of the population covariance matrix given the spectral of the sample covariance matrix.
#' @keywords population spectral moments estimators
#' @name moments_PSD
#' @param eigenvalues all eigenvalues of the sample covariance matrix including 0's.
#' @param n degree of freedom of the sample covariance matrix.
#' @param mom_degree the maximum order of moments.
#' @return Estimators of moments from the first to the \code{mom_degree} -th order.
#' @references  Bai, Z., Chen, J., & Yao, J. (2010).
#'\emph{On estimation of the population spectral distribution from a high-dimensional sample covariance matrix.}
#' Australian & New Zealand Journal of Statistics, 52(4), 423-437.
#' @examples
#' set.seed(10086)
#' n = 400; p= 500
#' pop_eig = seq(10,1,length = p)
#' # Data with covariance matrix diag(pop_eig)
#' Z = matrix(rnorm(n*p),n,p)
#' X = Z %*% diag(sqrt(pop_eig))
#' raw_eig = svd(cov(X))$d
#' emp_eig = raw_eig[raw_eig>=0]
#' # Moments of population spectral distribution
#' colMeans(outer(pop_eig, 1:4, "^"))
#' # Estimators
#' moments_PSD(emp_eig, n-1, 4)
#' @export
moments_PSD = function( eigenvalues,
                        n,
                        mom_degree){
        if(!is.vector(eigenvalues, mode = "numeric")){
                stop("eigenvalues must be a numeric vector")
        }else{
                if( any(eigenvalues<0)){
                        stop("eigenvalues must be nonnegative.")
                }else{
                        eigenvalues = sort(eigenvalues, decreasing = TRUE)
                        p = length(eigenvalues)
                }
        }
        if(!is.numeric(n)){
                stop("n must be a numeric atomic")
        }
        if(length(n)!= 1){
                stop("length of n should be 1.")
        }
        if(n<=0){
                stop("n must be positive.")
        }
        if(ceiling(n) != n){
                stop("n must be integer.")
        }
        if(!is.numeric(mom_degree)){
                stop("mom_degree must be numeric")
        }
        if(length(mom_degree)!= 1){
                stop("length of mom_degree should be 1.")
        }
        if(mom_degree <1){
                stop("mom_degree must be no less than 1.")
        }
        if(ceiling(mom_degree) != mom_degree){
                stop("mom_degree must be integer.")
        }
        gamma = p/n
        if(p >= n){
                emp_dual = eigenvalues[1:n]
        }else{
                emp_dual = c(eigenvalues, rep(0,n-p))
        }
        emp_moments = colMeans(outer( emp_dual, 1:mom_degree, FUN = '^'))
        pop_moments = numeric(mom_degree)
        pop_moments[1] = mean(eigenvalues)

        if(mom_degree > 1){
                # recursive formulas; cannot be parallel
                for( kk in 2:mom_degree){
                        max_values = kk %/% (1:(kk-1)) # The maximum possible value of i1,..., i_(kk-1) because j * i_{j}<= kk
                        possible_values = mapply(seq, 0, max_values, SIMPLIFY = FALSE) # Possible values of i
                        partitions = t(expand.grid(possible_values)) # all possible partitions
                        valid_partitions = partitions[, colSums(partitions * (1:(kk-1))) == kk, drop = FALSE] # valid partitions
                        # Coefficients
                        fac1 = (gamma^(colSums(valid_partitions)))  # gamma^(i1+i2+i3+...+ij)
                        fac2 = factorial(kk) / apply(factorial(valid_partitions), 2, prod) # fac/fac3 is the phi in equation (8) of the reference
                        fac3 = factorial(kk+1 - colSums(valid_partitions))
                        offset = sum( fac1 * fac2 /fac3 * apply( (pop_moments[1:(kk-1)])^valid_partitions, 2, prod ))
                        pop_moments[kk]= (emp_moments[kk] - offset) / gamma
                }
        }
        return(pop_moments)
}


R/eigen_proj_1samp.R
#' An internal function
#' @name eigen_proj_1samp
#' @keywords internal

eigen_proj_1samp = function(X,
                            mu_0,
                            lower_lambda = NULL){
        N = nrow(X)
        p = ncol(X)
        nn = N-1L
        #gamma = p/nn
        X_bar = colMeans(X)
        half_S = 1 / sqrt(nn) * t(X) %*%  (diag(1,nrow = N, ncol = N) - 1/N * matrix(1,N,N)) # S = half_S %*% t(half_S)

        svd_half_S = try(svd(half_S, nu = p ,  nv = 0), silent = TRUE)

        # Handle the situation where svd fails to converge
        if(inherits(svd_half_S,"try-error")){
                S = half_S %*% t(half_S)
                # If lower_lambda specified, add the lower bound to S, then svd.
                # If not specified, generate recommended lower bound.
                # Initially (mean(diag(S))/100), if not enough, ridge = 1.5 * ridge.
                if(!is.null(lower_lambda)){
                        ridge = lower_lambda
                        svdofS_ridge = try(svd(S + diag(ridge, nrow = p), nu = p , nv = 0), silent = TRUE)
                        if(inherits(svdofS_ridge, "try-error")){
                                stop("The lower bound of lambda sequence is too small.")
                        }
                }else{  # the generated lower bound of lambda sequence
                        ridge = (mean(diag(S))/100)
                        svdofS_ridge = try(svd(S + diag(ridge, nrow = p), nu = p , nv = 0), silent = TRUE)
                        loop_counter = 0L
                        while(inherits(svdofS_ridge, "try-error") && loop_counter<=20L){
                                ridge = ridge * 1.5
                                loop_counter = loop_counter +1L
                                svdofS_ridge = try(svd(S + diag(ridge, nrow = p), nu = p , nv = 0), silent = TRUE)
                        }
                        if(loop_counter > 20L)
                                stop("singular value algorithm in svd() did not converge")
                }
                # projection of X_bar - mu_0 to the eigenspace of S.
                emp_evec = svdofS_ridge$u
                # To speed up the computation of Stieltjes transform, separate positive eigenvalues and negative ones.
                emp_eig  = (svdofS_ridge$d - ridge) * (svdofS_ridge$d >= ridge)
                positive_emp_eig = emp_eig[emp_eig > (1e-8) * mean(emp_eig)]
                #num_zero_emp_eig = p - length(positive_emp_eig) # number of 0 eigenvalues
        }else{
                emp_evec = svd_half_S$u
                eig_raw = svd_half_S$d^2
                positive_emp_eig = eig_raw[eig_raw > (1e-8 ) * mean(eig_raw)]
                num_zero_emp_eig = p - length(positive_emp_eig)
                emp_eig = c(positive_emp_eig, rep(0, times = num_zero_emp_eig))
                if(!is.null(lower_lambda)){
                        ridge = lower_lambda
                }else{
                        ridge =  (mean(emp_eig)/100)
                }
        }
        proj = as.vector(sqrt(N) * t(emp_evec) %*% (X_bar - mu_0))
        return(list(n = nn, pos_eig_val = positive_emp_eig, proj_shift = proj, lower_lambda = ridge))
}




R/eigen_proj_2samp.R
#' An internal function
#' @name eigen_proj_2samp
#' @keywords internal

eigen_proj_2samp = function(X,
                            Y,
                            mu_0,
                            lower_lambda = NULL){
        #X = matrix(rnorm(100 * 50),100, 50)
        #Y = matrix(rnorm(200 * 50),200, 50)

        N1 = nrow(X)
        N2 = nrow(Y)
        p = ncol(X)
        nn = N1 + N2 - 2
        #gamma = p/nn

        X_bar = colMeans(X)
        Y_bar = colMeans(Y)
        Z = cbind(t(X),t(Y))
        design_mat = rbind(c(rep(1,times = N1), rep(0,times =N2)),
                           c(rep(0,times = N1),rep(1,times =N2)))
        half_S = 1/sqrt(nn) * Z %*% (diag(1, nrow = N1+N2 )  -
                                      t(design_mat) %*% diag(c(1/N1, 1/N2), nrow =2) %*% design_mat)
        svd_half_S = try(svd(half_S, nu = p , nv = 0), silent = TRUE)

        # Handle the situation where svd fails to converge
        if(inherits(svd_half_S,"try-error")){
                S = (cov(X) * (N1-1) + cov(Y) * (N2-1)) / nn    # or svd_half_S %*% t(svd_half_S)
                # If lower_lambda specified, add the lower bound to S, then svd.
                # If not specified, generate recommended lower bound.
                # Initially (mean(diag(S))/100), if not enough, ridge = 1.5 * ridge.
                if(!is.null(lower_lambda)){
                        ridge = lower_lambda
                        svdofS_ridge = try(svd(S + diag(ridge, nrow = p), nu = p , nv = 0), silent = TRUE)
                        if(inherits(svdofS_ridge, "try-error")){
                                stop("The lower bound of lambda sequence is too small.")
                        }
                }else{  # the generated lower bound of lambda sequence
                        ridge = (mean(diag(S))/100)
                        svdofS_ridge = try(svd(S + diag(ridge, nrow = p), nu = p, nv = 0), silent = TRUE)
                        loop_counter = 0
                        while(inherits(svdofS_ridge, "try-error") & loop_counter<=20){
                                ridge = ridge * 1.5
                                loop_counter = loop_counter +1
                                svdofS_ridge = try(svd(S + diag(ridge, nrow = p), nu = p, nv = 0), silent = TRUE)
                        }
                        if(loop_counter > 20)
                                stop("singular value algorithm in svd() did not converge")
                }
                # projection of X_bar - mu_0 to the eigenspace of S.
                emp_evec = svdofS_ridge$u
                # To speed up the computation of Stieltjes transform, separate positive eigenvalues and negative ones.
                emp_eig  = (svdofS_ridge$d - ridge) * (svdofS_ridge$d >= ridge)
                positive_emp_eig = emp_eig[emp_eig > 1e-8 * mean(emp_eig)]
               #num_zero_emp_eig = p - length(positive_emp_eig) # number of 0 eigenvalues
        }else{
                emp_evec = svd_half_S$u
                eig_raw = svd_half_S$d^2
                positive_emp_eig = eig_raw[eig_raw > (1e-8) * mean(eig_raw)]
                num_zero_emp_eig = p - length(positive_emp_eig)
                emp_eig = c(positive_emp_eig, rep(0, times = num_zero_emp_eig) )
                if(!is.null(lower_lambda)){
                        ridge = lower_lambda
                }else{
                        ridge =  (mean(emp_eig)/100)
                }
        }
        proj = as.vector(sqrt((N1*N2/(N1+N2))) * t(emp_evec) %*% (X_bar - Y_bar - mu_0))
        return(list(n = nn, pos_eig_val =positive_emp_eig, proj_shift = proj, lower_lambda = ridge))
}



R/ARHT.R
#' An adaptable generalized Hotelling's \eqn{T^2} test for high dimensional data
#' @export
#' @import stats
#' @description This function performs the adaptable regularized Hotelling's \eqn{T^2} test (ARHT) (Li et al., (2016) <arXiv:1609.08725>) for the one-sample
#'              and two-sample test problem, where we're interested in detecting the mean vector in the one-sample problem or the difference
#'              between mean vectors in the two-sample problem in a high dimensional regime.
#'
#' @details The method incorporates ridge-regularization in the classic Hotelling's \eqn{T^2} test with the regularization parameter
#'          chosen such that the asymptotic power under a class of probabilistic alternative prior models is maximized. ARHT combines
#'          different prior models by taking the maximum of statistics under all models. ARHT is distributed as the maximum
#'          of a correlated multivariate normal random vector. We estimate its covariance matrix and bootstrap its distribution. The
#'          returned p-value is a Monte Carlo approximation to its true value using the bootstrap sample, therefore not deterministic.
#'          Various methods are available to calibrate the slightly inflated Type 1 error rate of ARHT, including Cube-root transformation,
#'          square-root transformation and chi-square approximation.
#' @param X the n1-by-p observation matrix with numeric column variables.
#' @param Y an optional n2-by-p observation matrix; if \code{NULL}, a one-sample test is conducted on \code{X}; otherwise, a two-sample test
#'          is conducted on \code{X} and \code{Y}.
#' @param mu_0 the null hypothesis vector to be tested; if \code{NULL}, the default value is the 0 vector of length p.
#' @param prob_alt_prior a non-empty list; Each field is a numeric vector with sum 1. The default value is the "canonical weights"
#'        \code{list(c(1,0,0), c(0,1,0), c(0,0,1))}; Each field represents a probabilistic prior model specified by weights of \eqn{I_p},
#'        \eqn{\Sigma}, \eqn{\Sigma^2}, etc, where \eqn{\Sigma} is the population covariance matrix of the observations.
#' @param Type1error_calib the method to calibrate Type 1 error rate of ARHT. Choose its first element when more than one are specified.
#'        Four values are allowed:
#'        \itemize{\item{\code{cube_root}} The default value; cube-root transformation;
#'                 \item{\code{sqrt}} Square-root transformation;
#'                 \item{\code{chi_sq}} Chi-square approximation, not available when more than three models are specified in \code{prob_alt_prior};
#'                 \item{\code{none}} No calibration.
#' }
#' @param lambda_range optional user-supplied lambda range; If \code{NULL}, ARHT chooses its own range.
#' @param nlambda optional user-supplied number of lambda's in grid search; default to be \code{2000}; the grid is progressively coarser.
#' @param bs_size positive numeric with default value \code{1e5}; only effective when more than one prior models are specified in \code{prob_alt_prior};
#'        control the size of the bootstrap sample used to approximate the ARHT p-value.
#' @references Li, H. Aue, A., Paul, D. Peng, J., & Wang, P. (2016). \emph{An adaptable generalization of Hotelling's \eqn{T^2} test in high dimension.}
#'             <arXiv:1609:08725>.
#' @references Chen, L., Paul, D., Prentice, R., & Wang, P. (2011). \emph{A regularized Hotelling's \eqn{T^2} test for pathway analysis in proteomic studies.}
#'             Journal of the American Statistical Association, 106(496), 1345-1360.
#' @return \itemize{
#'  \item{\code{ARHT_pvalue}}: The p-value of ARHT test.
#'                             \itemize{
#'                              \item If \code{length(prob_alt_prior)==1}, it is identical to \code{RHT_pvalue}.
#'                              \item If \code{length(prob_alt_prior)>1}, it is the p-value after combining results from all prior models. The value is
#'                                    bootstrapped, therefore not deterministic.
#'                              }
#'  \item{\code{RHT_opt_lambda}}: The optimal lambda's chosen under each of the prior models in \code{prob_alt_prior}. It has the same length and order as
#'                              \code{prob_alt_prior}.
#'  \item{\code{RHT_pvalue}}: The p-value of RHT tests with the lambda's in \code{RHT_opt_lambda}.
#'  \item{\code{RHT_std}}: The standardized RHT statistics with the lambda's in \code{RHT_opt_lambda}.
#'  Take its maximum to get the statistic of ARHT test.
#'  \item{\code{Theta1}}: As defined in Li et al. (2016) <arXiv:1609.08725>, the estimated asymptotic means of RHT statistics with the lambda's in \code{RHT_opt_lambda}.
#'  \item{\code{Theta2}}: As defined in Li et al. (2016) <arXiv:1609.08725>, \code{2*Theta2} are the estimated asymptotic variances of RHT statistics the lambda's in \code{RHT_opt_lambda}.
#'  \item{\code{Corr_RHT}}: The estimated correlation matrix of the statistics in \code{RHT_std}.
#'}
#' @examples
#' set.seed(10086)
#' # One-sample test
#' n1 = 300; p =500
#' dataX = matrix(rnorm(n1 * p), nrow = n1, ncol = p)
#' res1 = ARHT(dataX)
#'
#' # Two-sample test
#' n2= 400
#' dataY = matrix(rnorm(n2 * p), nrow = n2, ncol = p )
#' res2 = ARHT(dataX, dataY, mu_0 = rep(0.01,p))
#'
#' # Specify probabilistic alternative priors model
#' res3 = ARHT(dataX, dataY, mu_0 = rep(0.01,p),
#'      prob_alt_prior = list(c(1/3, 1/3, 1/3), c(0,1,0)))
#'
#' # Change Type 1 error calibration method
#' res4 = ARHT(dataX, dataY, mu_0 = rep(0.01,p),
#'      Type1error_calib = "sqrt")
#'
#' RejectOrNot = res4$ARHT_pvalue < 0.05
#'

ARHT = function(X,
                Y = NULL,
                mu_0 = NULL,
                prob_alt_prior = list(c(1, 0, 0), c(0, 1, 0), c(0, 0, 1)),
                Type1error_calib = c("cube_root", "sqrt", "chi_sq", "none"),
                lambda_range = NULL,
                nlambda = 2000,
                bs_size = 1e5){

        if(length(dim(X)) > 2L || !(is.numeric(X)))
                stop("X must be a numeric matrix with column variables")
        if(!is.matrix(X))
                X = as.matrix(X)
        if(nrow(X) <= 1L){
                stop("The number of rows in X must be larger than 1")
        }

        if(is.null(Y)){
                mode = "one_sample"
        }else{
                if(length(dim(Y)) > 2L || !(is.numeric(Y)))
                        stop("Y must be a numeric matrix with column variables")
                if(!is.matrix(Y))
                        Y <- as.matrix(Y)
                if(nrow(Y) <= 1L){
                        stop("The number of rows in Y must be larger than 1")
                }
                if(ncol(Y) != ncol(X)){
                        stop("The dimensions of X and Y differ")
                }
                mode = "two_sample"
        }

        if(!is.null(mu_0)){
                if(!is.vector(mu_0, mode = "numeric"))
                        stop("mu_0 must be a numeric vector")
                if(length(mu_0) != ncol(X))
                        stop("The dimension of X doesn't match with that of mu_0")
        }else{
                mu_0 = numeric(ncol(X))
        }

        if(!is.list(prob_alt_prior))
                stop("prob_alt_prior must be a list of numeric vectors")

        if(!all(sapply(prob_alt_prior, is.vector, mode = "numeric")))
                stop("prob_alt_prior must be a list of numeric vectors")

        valid_prob_alt_prior = sapply(prob_alt_prior, function(a){ round(sum(a), 5) != 1})
        if(any(valid_prob_alt_prior)){
                stop(paste("In Model", paste( which(valid_prob_alt_prior), collapse = ", "),
                           "specified in prob_alt_prior, the sum of prior weights is not 1"))
        }

        # throw away meaningless 0's and shorten prob_alt_prior
        max_nonzero_index = max(sapply(prob_alt_prior, function(xxx) max(which(xxx != 0))))
        prob_alt_prior = lapply(prob_alt_prior, function(xxx){
                xxx[1: min(length(xxx), max_nonzero_index)]})

        if(!is.null(lambda_range)){
                if(!is.vector(lambda_range, mode = "numeric"))
                        stop("lambda_range must be a numeric vector of two elements")
                if(length(lambda_range)!=2L)
                        stop("The length of lambda_range must be 2.")
                if(lambda_range[1]<=0)
                        stop("The lower bound of lambda sequence must be positive")
                if(lambda_range[2]<= lambda_range[1])
                        stop("The upper bound of lambda sequence must be larger than the lower bound")
        }

        if( (!is.numeric(nlambda)) || (length(nlambda)!= 1) )
                stop("nlambda must be numeric of length 1")
        if(nlambda<=0)
                stop("nlambda must be postive")
        nlambda = ceiling(nlambda)

        if(!(Type1error_calib[1] %in% c("cube_root", "sqrt", "chi_sq", "none"))){
                Type1error_calib = "cube_root"
                warning('Unknown value for Type1error_calib; default value "cube_root" is chosen instead')
        }
        if( (length(prob_alt_prior) >3L) && (Type1error_calib[1] == "chi_sq")){
                stop("Chi-square calibration of Type 1 error is not available when the number of prior models in prob_alt_prior
                     is larger than 3")
        }
        if(length(prob_alt_prior)>1L ){
                if( (!is.numeric(bs_size)) || (length(bs_size)!= 1) )
                        stop("bs_size must be numeric of length 1")
                if(bs_size <=0 )
                        stop("bs_size must be postive")
                if(bs_size < 1e3)
                        warning("Bootstrap sample size is too small; ARHT_pvalue is not reliable.")
        }
        bs_size = ceiling(bs_size)

        if((Type1error_calib[1] != "chi_sq") && (length(prob_alt_prior)>1L)){
                bootstrap_sample = matrix(rnorm(length(prob_alt_prior)*bs_size),
                                          ncol = bs_size)
        }

        if(mode == "one_sample"){
                eig_proj = eigen_proj_1samp(X, mu_0, lower_lambda = lambda_range[1])
        }else{
                eig_proj = eigen_proj_2samp(X, Y, mu_0, lower_lambda = lambda_range[1])
        }
        p = ncol(X)
        n = eig_proj$n
        gamma = p/n
        proj_diff = eig_proj$proj_shift
        ridge = eig_proj$lower_lambda

        # To speed up the computation of Stieltjes transform, separate positive eigenvalues and negative ones.
        positive_emp_eig = eig_proj$pos_eig_val
        num_zero_emp_eig = p - length(positive_emp_eig)
        emp_eig = c(positive_emp_eig, rep(0, times = num_zero_emp_eig))

        ## specify the lambda's net. Use log-scale. Progressively coarser.
        if(is.null(lambda_range)){
                lambda = exp(seq(from = log(ridge),
                                 to = log(20 * emp_eig[1] + (ridge - mean(emp_eig)/100) * (ridge - mean(emp_eig)/100 >0)),
                                 length = nlambda))
        }else{
                lambda = exp(seq(from = log(lambda_range[1]),
                                 to = log(lambda_range[2]),
                                 length = nlambda))
        }

        ## Stieltjes transform, its derivative, Theta_1, Theta_2
        mF = 1/p * ( rowSums(1/outer(lambda, positive_emp_eig, FUN = "+"))
                     + num_zero_emp_eig/lambda )

        mFprime = 1/p * (rowSums(1/(outer(lambda, positive_emp_eig, FUN = "+"))^2)
                         + num_zero_emp_eig/lambda^2)
        Theta1 = (1 - lambda*mF)/(1 - gamma*(1 - lambda * mF))

        Theta2 = (1 + gamma*Theta1)^2 * (Theta1 - lambda *(mF - lambda * mFprime)/(1 - gamma*(1 - lambda * mF))^2)

        # Calculate the power under each prior model
        prior_max_order = max(sapply(prob_alt_prior,length))
        unified_prob_alt_prior = lapply(prob_alt_prior, function(i) c(i, rep(0, times = max(prior_max_order,2) - length(i))))
        matrix_prob_alt_prior = do.call(rbind, unified_prob_alt_prior)
        if(prior_max_order <= 2L){
                rhos = rbind(mF, Theta1)
        }else{
                pop_moments = moments_PSD(emp_eig, n, prior_max_order-2)
                rhos = matrix(NA, nrow = prior_max_order, ncol = length(mF))
                rhos[1, ] = mF
                rhos[2, ] = Theta1
                # recursive formulas; cannot be parallel
                for(ii in 3:prior_max_order){
                        rhos[ii,] = (1 + gamma * Theta1) * (pop_moments[ii-2] - lambda * rhos[ii-1, ])
                }
        }
        powers = t(matrix_prob_alt_prior %*% rhos) / sqrt(2*gamma*Theta2) # Column: prior model; Row: lambda

        opt_lambda_index = apply(powers, 2, which.max) # optimal lambda index under each prior model

        ## Estimated covariance matrix of standardized RHT statistics with optimal lambda's
        G = matrix( apply( expand.grid(opt_lambda_index, opt_lambda_index), 1,
                           function(ddd){
                                aaa = ddd[1]
                                bbb = ddd[2]
                                if( abs(aaa - bbb) < 1e-8){
                                        return(1)
                                }else{
                                        return( (1 + gamma * Theta1[aaa]) * (1 + gamma * Theta1[bbb]) * (
                                                lambda[aaa] * Theta1[aaa] - lambda[bbb] * Theta1[bbb]) / (
                                                        (lambda[aaa] - lambda[bbb]) * sqrt(Theta2[aaa] * Theta2[bbb]))
                                                )
                                }
                                }),
                    nrow = length(opt_lambda_index), ncol = length(opt_lambda_index) )

        ## square root of G ##
        G_eigen = eigen(G,symmetric=T) ### project G to the ''closest'' nonnegative definite matrix
        G_evec = G_eigen$vectors
        G_eval = G_eigen$values
        G_eval_plus = G_eval * (G_eval >= 0)
        G_sqrt = G_evec %*% diag(sqrt(G_eval_plus))

        # standardized statistics
        RHT = sapply(lambda[opt_lambda_index], function(xx){
                (1/p) * sum( proj_diff^2 / (emp_eig + xx))}
        )
        if(Type1error_calib[1] != "chi_sq"){
                if(Type1error_calib[1] == "cube_root"){
                        RHT_std = {sqrt(p) * ( RHT^(1/3) - (Theta1[opt_lambda_index])^(1/3)) /
                                        sqrt(2*Theta2[opt_lambda_index]) / (1 / 3 * Theta1[opt_lambda_index]^(-2/3))}
                }
                if(Type1error_calib[1] == "sqrt"){
                        RHT_std = {sqrt(p) * (sqrt(RHT) - sqrt(Theta1[opt_lambda_index]))/
                                        sqrt(Theta2[opt_lambda_index] / 2 / Theta1[opt_lambda_index])}
                }
                if(Type1error_calib[1] == "none"){
                        RHT_std = (RHT - Theta1[opt_lambda_index]) / sqrt(2 * Theta2[opt_lambda_index] / p)
                }
                # p-values
                if(length(prob_alt_prior) == 1){
                        p_value = 1 - pnorm(RHT_std)
                        composite_p_value = p_value
                }else{
                        p_value = 1 - pnorm(RHT_std)
                        Tmax = apply(G_sqrt %*% bootstrap_sample,2,max)
                        composite_p_value = 1 - mean(max(RHT_std)>Tmax)
                }
        }

        if(Type1error_calib[1] == "chi_sq"){
                if(length(prob_alt_prior) == 1L){
                        # when one prior model is specified, no need for bootstrap
                        constant_coef = Theta2[opt_lambda_index] / Theta1[opt_lambda_index]
                        degree_freedom =  p * (Theta1[opt_lambda_index])^2 / Theta2[opt_lambda_index]
                        p_value = 1 - pchisq( p * RHT / constant_coef, df = degree_freedom)
                        composite_p_value = p_value
                }else{
                        if( length(prob_alt_prior) == 2L){
                                # Trick: add dummy variables to make the length of opt_lambda_index when less than 3 priors are specified
                                # max(RHT(lambda_1), RHT(lambda_2), RHT(lambda_1)) = max(RHT(lambda_1), RHT(lambda_2))
                                length3_opt_lambda_index = c(opt_lambda_index, opt_lambda_index[1])
                                # expand G to 3 variables
                                G_tmp = G_sqrt %*% t(G_sqrt)
                                G_expand = rbind( cbind(G_tmp, c(1, G_tmp[1,2])), c(1, G_tmp[1,2], 1))
                        }else{  # when 3 prior models are specified
                                length3_opt_lambda_index = opt_lambda_index
                                G_expand = G_sqrt %*% t(G_sqrt)
                        }
                        # Call r3chisq to get the generated 3-vairate chi-square bootstrap sample
                        constant_coef = Theta2[length3_opt_lambda_index] / Theta1[length3_opt_lambda_index]
                        degree_freedom = ceiling( p * (Theta1[length3_opt_lambda_index])^2 / Theta2[length3_opt_lambda_index])
                        chisq = r3chisq(size = bs_size, df = degree_freedom, corr_mat = G_expand)$sample
                        # standardize the bootstrap sample
                        T1sample = (1/p*constant_coef[1] * chisq[,1] - Theta1[opt_lambda_index[1]])/sqrt(2*Theta2[opt_lambda_index[1]]/p)
                        T2sample = (1/p*constant_coef[2] * chisq[,2] - Theta1[opt_lambda_index[2]])/sqrt(2*Theta2[opt_lambda_index[2]]/p)
                        T3sample = (1/p*constant_coef[3] * chisq[,3] - Theta1[opt_lambda_index[3]])/sqrt(2*Theta2[opt_lambda_index[3]]/p)
                        Tmax = pmax(T1sample, T2sample, T3sample)
                        p_value = 1 - pchisq( p * RHT / constant_coef[1:length(RHT)], df = degree_freedom[1:length(RHT)])
                        RHT_std = (RHT - Theta1[opt_lambda_index])/sqrt(2*Theta2[opt_lambda_index]/p)
                        composite_p_value = 1 - mean(max(RHT_std) > Tmax)
                }
        }
        return(list(ARHT_pvalue = composite_p_value,
                    RHT_pvalue = p_value,
                    RHT_std = RHT_std,
                    RHT_opt_lambda = lambda[opt_lambda_index],
                    Theta1 = Theta1[opt_lambda_index],
                    Theta2 = Theta2[opt_lambda_index],
                    Corr_RHT = G
                    ))
}


